File: src/core/net/strategies/switch-account.ts
import type { NgaCredentials } from '../auth'
import { DEFAULT_ROTATION_FORMATS, interfaceKeyOf, type FetchCombo } from '../combo'
import { NgaError } from '../errors'
import type { FetchContext, FetchStrategy, NgaRequest, StrategyOutcome } from '../types'
import { runAttempt } from './attempt'

export const SWITCH_ACCOUNT_STRATEGY_NAME = 'switch-account'

export interface SwitchAccountOptions {
  /**
   * 已登录账号的凭证表，顺序即账号管理页的顺序。
   *
   * 注入而不是 import store：core 层零 RN 依赖，账号存在 SecureStore 里，
   * 拿到这一层的只能是纯数据（src/store/accounts.ts 提供实现）。
   */
  readonly listCredentials: () => readonly NgaCredentials[]
}

/**
 * 当前凭证之后的下一个账号（循环取）。
 * 只有一个账号（或一个都没有）时返回 null——没得换，这一档就不该启用。
 * 当前凭证不在表里（刚退出登录之类）时从头一个开始。
 */
export function nextCredentialsAfter(
  accounts: readonly NgaCredentials[],
  current: NgaCredentials | null,
): NgaCredentials | null {
  if (accounts.length < 2) return null
  const index = current === null ? -1 : accounts.findIndex((item) => item.uid === current.uid)
  return accounts[(index + 1) % accounts.length] ?? null
}

/**
 * 反封锁链的换账号重试档（ADR-0002；Android v4 的对策）：
 * 取下一个已登录账号的 cookie **只试一次**，成了就成了，不成交给链上后面的兜底。
 *
 * 只在多账号时启用：单账号换来换去还是同一个 cookie，白等一次往返。
 * 用哪个组合不再枚举——前一档 format-rotation 已经把组合空间跑完了，
 * 这一档变的是**身份**，所以沿用缓存里的组合（没有就用默认档）。
 */
export function createSwitchAccountStrategy(options: SwitchAccountOptions): FetchStrategy {
  return {
    name: SWITCH_ACCOUNT_STRATEGY_NAME,
    async run(request: NgaRequest, context: FetchContext): Promise<StrategyOutcome> {
      const via = SWITCH_ACCOUNT_STRATEGY_NAME
      const current =
        request.credentials === undefined ? context.credentials : request.credentials
      const next = nextCredentialsAfter(options.listCredentials(), current)
      if (next === null) {
        return {
          ok: false,
          error: new NgaError({
            kind: 'unavailable',
            message: '只有一个已登录账号，没有可换的',
            via,
          }),
        }
      }

      const cached = context.comboCache?.get(interfaceKeyOf(request))
      const combo: FetchCombo = cached ?? {
        format: request.format ?? DEFAULT_ROTATION_FORMATS[0] ?? 'json',
        host: request.host ?? context.host,
      }
      return runAttempt(request, context, { via, combo, credentials: next })
    },
  }
}

File: app/src/main/kotlin/com/chasel/ng2n/core/net/strategies/SwitchAccount.kt
package com.chasel.ng2n.core.net.strategies

import com.chasel.ng2n.core.net.AccountPolicy
import com.chasel.ng2n.core.net.Credential
import com.chasel.ng2n.core.net.CredentialOverride
import com.chasel.ng2n.core.net.DEFAULT_ROTATION_FORMATS
import com.chasel.ng2n.core.net.FetchCombo
import com.chasel.ng2n.core.net.FetchContext
import com.chasel.ng2n.core.net.FetchStrategy
import com.chasel.ng2n.core.net.NgaError
import com.chasel.ng2n.core.net.NgaErrorKind
import com.chasel.ng2n.core.net.NgaRequest
import com.chasel.ng2n.core.net.ResponseFormat
import com.chasel.ng2n.core.net.StrategyOutcome
import com.chasel.ng2n.core.net.interfaceKeyOf

const val SWITCH_ACCOUNT_STRATEGY_NAME = "switch-account"

/**
 * 当前凭证之后的下一个账号(循环取)。直译 `nextCredentialsAfter`。
 *
 * 只有一个账号(或一个都没有)时返回 null —— 没得换,这一档就不该启用。
 * 当前凭证不在表里(刚退出登录之类)时从头一个开始。
 */
fun nextCredentialAfter(accounts: List<Credential>, current: Credential?): Credential? {
  if (accounts.size < 2) return null
  val index = if (current == null) -1 else accounts.indexOfFirst { it.uid == current.uid }
  return accounts.getOrNull((index + 1) % accounts.size)
}

/**
 * 反封锁链的换账号重试档(ADR-0002;Android v4 的对策)。
 * 直译 `src/core/net/strategies/switch-account.ts`。
 *
 * 取下一个已登录账号的 cookie **只试一次**,成了就成了,不成交给链上后面的兜底。
 *
 * 只在多账号时启用:单账号换来换去还是同一个 cookie,白等一次往返。
 * 用哪个组合不再枚举——前一档 format-rotation 已经把组合空间跑完了,
 * 这一档变的是**身份**,所以沿用缓存里的组合(没有就用默认档)。
 *
 * @param listCredentials 已登录账号的凭证表,顺序即账号管理页的顺序。
 *   注入而不是直接读仓库:core 层零 Android 依赖,凭证存在 Keystore 里,
 *   拿到这一层的只能是纯数据(`data/account/AccountStore` 提供实现)。
 */
class SwitchAccountStrategy(
  private val listCredentials: suspend () -> List<Credential>,
) : FetchStrategy {

  override val name: String = SWITCH_ACCOUNT_STRATEGY_NAME

  override suspend fun run(request: NgaRequest, context: FetchContext): StrategyOutcome {
    // ── 修 P1-01:写操作 / 钉死身份的请求禁入这一档 ──────────────────────────
    // 换账号重发一个非幂等写 = 把写入落到**别人**头上(审计里的「跨账号写入」)。
    // 装配上写请求走的是只有 direct 的那条链,这道闸是第二层保险。
    if (request.isWrite || request.accountPolicy == AccountPolicy.PINNED) {
      return unavailable("这条请求的身份已钉死(P1-01),不换账号重试")
    }

    val current = request.credential?.credential ?: context.credential
    val next = nextCredentialAfter(listCredentials(), current)
      ?: return unavailable("只有一个已登录账号,没有可换的")

    val cached = context.comboCache?.get(interfaceKeyOf(request))
    val combo = cached ?: FetchCombo(
      format = request.format ?: DEFAULT_ROTATION_FORMATS.firstOrNull() ?: ResponseFormat.JSON,
      host = request.host ?: context.host,
    )
    return runAttempt(
      request = request,
      context = context,
      via = SWITCH_ACCOUNT_STRATEGY_NAME,
      combo = combo,
      credentialOverride = CredentialOverride(next),
    )
  }

  private fun unavailable(message: String): StrategyOutcome = StrategyOutcome.Failed(
    NgaError(NgaErrorKind.UNAVAILABLE, message, via = SWITCH_ACCOUNT_STRATEGY_NAME),
  )
}
